### PR TITLE
add varargs version of ccall to preamble.js

### PR DESCRIPTION
**Pull request by ericmandel**
Add varargs version of ccall to preamble.js

For emscripten-discuss discussion leading up to this PR, see:

"problems making a ccall to a varargs routine (v1.37.20)"

https://groups.google.com/forum/#!searchin/emscripten-discuss/varargs%7Csort:relevance/emscripten-discuss/ZBiXvH-Eq1E/kzkTovDHAAAJ

### CODE REVIEW

**Comment by kripken** [comment]
Do we need a new function here? Couldn't this be an extra param to the current method?

**Comment by ericmandel** [comment]
An extension of the current call was the first implementation we looked at in the discussion above. It was jj's suggestion to split off into a separate call:
````
I think we'd probably want to flag the
presence of varargs with a separate function signature altogether,
such as Module.ccall_vararg(), to avoid any overhead to non-vararg
function calls (Module.cwrap() and Module.ccall() can be very on a
performance sensitive path). 
````
I'm agnostic about this question, happy to do it either way.

A separate issue perhaps worth mentioning at the start: we also came up with a string-like specifier for the varargs spec. But the repeated parsing of the text specification turned out to be too inefficient  for my needs (running thousands of calls over an image array), hence the alternative of passing an array spec directly. The latter works well for me, so a decision has to be made about the former (ease of specification, at the expense of efficiency, might be good for a single call).

File: src/preamble.js [hunk 1]
```diff
@@ -203,6 +203,183 @@ var cwrap, ccall;
...
+  // where cn specifiers are "s","i","u","d", as above
+  //
+  // # pass string, followed by a repeating series of double,int pairs:
+  // Module.ccall_varargs("miniprintf", "null", ["string", [["s"], ["d", "i"]]], ["%s %f %d %s %f %d %s\n", "foo", 1.234, 2, "foo1", 3.14, -100, "goo1"])
+  //
+  // The second form (the compiled version of the first form) can be used
+  // when ccall_varargs() is called repeatedly with the same varargs spec,
+  // e.g. call_varargs() over all elements in a large 2D array.
+  //
+  ccall_varargs = function ccallVarargsFunc(ident, returnType, argTypes, args, opts) {
```
Comment (nazar-pc) [inline src/preamble.js]: Looks like this should be an anonymous function, namely `ccall_varargs = function (ident, returnType, argTypes, args, opts) {`

**Comment by nazar-pc** [comment]
Would be nice to fix indentation and spacing everywhere, it is very difficult to read currently as alignment is misleading.

File: src/preamble.js (hunk 1 above)
Comment (ericmandel) [reply src/preamble.js]: The use of a function name follows the use of a similar name in the declaration of ccall().

Comment (nazar-pc) [reply src/preamble.js]: I mean, you never use `ccallVarargsFunc`, so I think you don't need it. There is `ccall_varargs` referencing the function already.

Comment (ericmandel) [reply src/preamble.js]: Understood ... the same situation exists with ccall():
````
ccall = function ccallFunc(ident, returnType, argTypes, args, opts) {
````
I assumed it was added for debugging/profiling, as described by kangax here:

````
https://kangax.github.io/nfe/
````

I have no problem removing it, but if I was the maintainer, I would prefer it match the ccall() declaration one way or another. (Note it's still unclear whether ccall_varargs should be a separate call from call() at all -- see Alon's comment above.)

Comment (nazar-pc) [reply src/preamble.js]: Got it. Closure Compiler will likely remove it anyway.

**Comment by kripken** [comment]
cc @juj who was active in the original discussion here

**Comment by ericmandel** [comment]
@kripken @juj status update? I am scheduled to give a talk in mid-Dec re: astrophysics data analysis in the browser and the punch line involves this call intimately. I probably can't do that without some hint of progress on this PR. If it has been rejected by neglect, can I please get advice about re-working the code?

**Comment by kripken** [comment]
Seems like @juj is busy, and I'm not sure I understand the discussion that preceded this PR, but here is how I see things:

 * I'm not sure why a new function is needed. Intuitively, I would guess we just need to add params to ccall, one for a list of vararg types and one for their values. Then all ccall does is allocate stack space for the varargs param and place those items on the stack - should be just a few lines. So the PR seems more complicated than I'd expect and I don't see why. Perhaps summarizing the approach taken here could help.
 * This is a fairly niche use case, so we probably shouldn't add it to the list of exported runtime methods by default. That way, it would be extra code in the majority of cases when it isn't used.
 * We will need tests for new functionality like this. They also help in reviewing.

**Comment by ericmandel** [comment]
A major complicating factor: how to specify the argument types when calling ccall_varargs(), if the number of passed arguments is not known until the call is made. Consider the following made-up routine:
````
double foo(double x, double y, ...)
````
Two example invocations showing different number of val/err pairs being supplied after two variable string args:
````
foo(512.0, 512.0, "pi", "err", 5, 0.5, 8, 0.6, -1, -1)
foo(512.0, 512.0, "pha", "err", 5, 0.5, 8, 0.6, 1, 0.3, 3, 0.26, -1, -1)
````
Since the end of the array is not known, the underlying code might do something like this:
````
  va_list args;
  va_start(args, y);
  s1 = va_arg(args, char *);
  s2 = va_arg(args, char *);
  while( 1 ){
     nval = va_arg(args, int);
     if( nval < 0 ){
          break;
      }
      nerr = va_arg(args, double);
     // process nval, nerr ....
  }
````
(See https://www.gnu.org/software/libc/manual/html_node/How-Many-Arguments.html#How-Many-Arguments for further discussion of end markers.)

So in this case there are two non-repeating string args, followed by a series of int/double pairs, with the end of the argument list marked by a negative int. The problem is how to specify the data types of the variable args.

I'm not sure how to do it using the method you mention, without forcing the javascript that calls ccall_varargs() to count the args and generate the data types.  Instead, I proposed this:
````
foo: function(){
    return Module.ccall_varargs('foo', 'number',
				[ 'number', 'number', [['s','s'],['i','d']]], Array.prototype.slice.call(arguments));
},
````
with the ccall_varargs implementation packaging up the repeated args as needed. 

A secondary complication comes from @juj 's suggestion to pass regexp-type strings to specify the non-repeating and repeating args, instead of the two arrays I used above. I found parsing these strings into arrays to be too slow for our analysis needs, but left the implementation intact for discussion.

Agreed that this might not be appropriate for the default list of exported methods.

To test, I compiled a simplified version of printf, which allows tests such as this:
````
  function test(){
    var i;
    for(i=0; i<3; i++){
        // string-style data type specification
        Module.ccall_varargs("miniprintf", "null", ["string", "[si]"], ["Hello %s: %d\n", "world", i]);
        Module.ccall_varargs("miniprintf", "null", ["string", "[sid]"], ["Hello %s: %d %f\n", "world", i, i + (i/10)]);
        Module.ccall_varargs("miniprintf", "null", ["string", "[si(dis)+]"], ["%s %d: %f %d %s %f %d %s\n", "string", i, 1.234 + i, 2, "foo1", 3.14159, -100, "goo1"]);

        // direct array specification of data types
        Module.ccall_varargs("miniprintf", "null", ["string", [["s", "i"], ["d", "i", "s"]]], ["%s %d: %f %d %s %f %d %s\n", "array ", i, 1.234 + i, 2, "foo1", 3.14159, -100, "goo1"]);
    }
  }
````

**Comment by juj** [comment]
This is a great example of a feature that may be super-useful to some, but where we should now diverge from the original method that `ccall` was linked in.

Not all users need `ccall` to be able to handle varargs, which is why I'm thinking this would be good to be a separate function such as `ccall_varargs`, rather than adding all the logic of varargs handling to the existing `ccall`.

I think that we should push `ccall`, `cwrap` and this new `ccall_varargs` out to a library file such as a `src/library_ccall.js`, and require users to pass `-s EXTRA_EXPORTED_RUNTIME_METHODS=['ccall']` or `-s EXTRA_EXPORTED_RUNTIME_METHODS=['ccall_varargs']`when they need this feature. That way it will be available in "core distribution", but users don't get it compiled in by default, and when they want to use it, they specify it to be exported.

I think this is the MO that we should approach pretty much all of our core runtime functions that users potentially call from external JS that there is no automatic mechanism to know if the item can be DCEd or not.



> Instead, I proposed this:

> [ 'number', 'number', [['s','s'],['i','d']]]

This looks fine, though could we do 'ss' and 'id' as flat strings instead of arrays of characters, since that should have much better JS GC footprint?

> A secondary complication comes from @juj 's suggestion to pass regexp-type strings to specify the non-repeating and repeating args, instead of the two arrays I used above. I found parsing these strings into arrays to be too slow for our analysis needs, but left the implementation intact for discussion.

No need to complicate this - if the regexp method is not good, let's just throw it out as unsupported.

**Comment by ericmandel** [comment]
> This looks fine, though could we do 'ss' and 'id' as flat strings instead of arrays of characters, since that should have much better JS GC footprint?

Good idea, I'll do that in the next update, assuming we get that far.

> No need to complicate this - if the regexp method is not good, let's just throw it out as unsupported.

As well as remove this too-slow code.

As well as try to add some tests ... probably have to re-learn Python first ...

**Comment by ericmandel** [comment]
updated calling sequence using flat strings is indeed nicer:
````
        // canonical case: string/int/double with no repeating arguments
        Module.ccall_varargs("miniprintf", "null", ["string", ["sid"]], ["Hello %s: %d %f\n", "world", i, i + (i/10)]);
        // same as above with explicitly empty repeat list
        Module.ccall_varargs("miniprintf", "null", ["string", ["sid", ""]], ["Hello %s: %d %f\n", "world", i, i + (i/10)]);
        // non-repeating string/int, then repeating double/int/string triplets
        Module.ccall_varargs("miniprintf", "null", ["string", ["si", "dis"]], ["%s %d:      %f %d %s %f %d %s\n", "iter ", i, 1.234 + i, 2, "foo1", 3.14159, -100, "goo1"]);
        // no non-repeating args, then repeating double/int/string triplets
        Module.ccall_varargs("miniprintf", "null", ["string", ["", "dis"]], ["same as above %f %d %s %f %d %s\n", 1.234 + i, 2, "foo1", 3.14159, -100, "goo1"]);
````

**Comment by kripken** [comment]
Ok, I think we can move forward here, without the regexp stuff, and not exporting this by default, as suggested by @juj. Perhaps we could put it in a JS library which would be even better.

About this example:
````
foo: function(){
    return Module.ccall_varargs('foo', 'number',
				[ 'number', 'number', [['s','s'],['i','d']]], Array.prototype.slice.call(arguments));
},
````
What does `[['s','s'],['i','d']]` mean?

**Comment by ericmandel** [comment]
It's the old syntax, before @juj suggested using flattened strings. The new syntax (PR was updated last week):
````
Module.ccall_varargs('foo', 'number',
				[ 'number', 'number', ['ss', 'id'], Array.prototype.slice.call(arguments));
````
which means:
There are two declared arguments in the routine, both of which are numbers.
These are followed by two non-repeating varargs, which are both strings.
These are followed by an unknown number of repeating arguments, all of which consist of an int followed by a double.
See example above "updated calling sequence using flat strings is indeed nicer", which shows:
````
Module.ccall_varargs("miniprintf", "null", ["string", ["si", "dis"]], ["%s %d:      %f %d %s %f %d %s\n", "iter ", i, 1.234 + i, 2, "foo1", 3.14159, -100, "goo1"]);
````
The 'string' is the non-varargs format statement, followed by varargs: a non-repeating string, then a non-repeating int, followed by two sets of repeating arguments (double, int, string).

**Pull request closed without merging**